### PR TITLE
feat: destroy deployment

### PR DESCRIPTION
**Pull request by tonsV2**
This PR can be tested using the following commands

    ./create-deployment.sh whoami my-deployment-1 some test deployment
    # Update the below id to match the output of the above command
    ./create-deployment-instance-dhis2-db.sh 1
    ./create-deployment-instance-dhis2-core.sh 1
    ./findDeploymentById.sh 1
    ./destroy-deployment.sh 1

### CODE REVIEW

File: pkg/instance/repository.go [hunk 1]
```diff
@@ -21,6 +23,34 @@ type repository struct {
...
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return errdef.NewNotFound("deployment not found by id: %d", id)
+		}
+		return fmt.Errorf("failed to delete deployment: %v", err)
+	}
+
+	for _, instance := range deployment.Instances {
+		err := r.db.Unscoped().Delete(&model.DeploymentInstance{}, instance.ID).Error
+		if err != nil {
+			return fmt.Errorf("failed to delete instance: %v", err)
```
Comment (teleivo) [inline pkg/instance/repository.go:38]:
```suggestion
			return fmt.Errorf("failed to delete instance %q: %v", instance.Name, err)
```

File: pkg/instance/repository.go [hunk 2]
```diff
@@ -21,6 +23,34 @@ type repository struct {
...
+		}
+	}
+
+	// TODO: The below is suppose to delete the instances as well but I can't make it work. So the above is needed...
```
Comment (teleivo) [inline pkg/instance/repository.go:42]: Please remove this if it does not work.

File: pkg/instance/service.go [hunk 3]
```diff
@@ -267,6 +269,7 @@ func (s service) DeployDeployment(token string, deployment *model.Deployment) er
 		return err
 	}
 
+	// TODO: Delete?
```
Comment (teleivo) [inline pkg/instance/service.go:272]: Why did you add this TODO?

File: pkg/instance/service.go [hunk 4]
```diff
@@ -313,6 +316,71 @@ func (s service) deployDeploymentInstance(token string, instance *model.Deployme
 	return nil
 }
 
+func (s service) DeleteDeployment(id uint) error {
+	deployment, err := s.FindDeploymentById(id)
```
Comment (teleivo) [inline pkg/instance/service.go:320]: We find the deployment 3 times if I am counting correctly 😅 Since we call `s.instanceRepository.DeleteDeployment(id)` only from here, can we pass in the `Deployment` to not fetch it again?

File: pkg/instance/service.go [hunk 5]
```diff
@@ -313,6 +316,71 @@ func (s service) deployDeploymentInstance(token string, instance *model.Deployme
...
+	deploymentGraph, err := s.validateNoCycles(deployment.Instances)
+	if err != nil {
+		return err
+	}
+
+	instances, err := deploymentOrder(deployment, deploymentGraph)
+	if err != nil {
+		return err
+	}
+
```
Comment (teleivo) [inline pkg/instance/service.go:334]: Please remove the newline as `deploymentOrder` and `Reverse` are closely related. Or best replace it with a comment of why you are calling `Reverse`. This might not be obvious at first.

File: pkg/instance/service.go [hunk 6]
```diff
@@ -313,6 +316,71 @@ func (s service) deployDeploymentInstance(token string, instance *model.Deployme
...
+	if errs != nil {
+		return errs
+	}
+
+	return s.instanceRepository.DeleteDeployment(id)
+}
+
+func (s service) destroyDeploymentInstance(instance *model.DeploymentInstance) error {
+	if instance.DeployLog != "" {
```
Comment (teleivo) [inline pkg/instance/service.go:353]: please invert this condition to remove nesting.

File: pkg/instance/handler.go [hunk 7]
```diff
@@ -1088,3 +1088,53 @@ func (h Handler) ListPublicInstances(c *gin.Context) {
...
+		return
+	}
+
+	user, err := handler.GetUserFromContext(c)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+
+	deployment, err := h.instanceService.FindDeploymentById(id)
```
[the remark this answers is not in the document]
Comment (tonsV2) [reply pkg/instance/handler.go:1120]: yes

File: pkg/instance/repository.go [hunk 8]
```diff
@@ -21,6 +21,25 @@ type repository struct {
 	instanceParameterEncryptionKey string
 }
 
+func (r repository) DeleteDeployment(deployment *model.Deployment) error {
+	for _, instance := range deployment.Instances {
+		err := r.db.Unscoped().Delete(&model.DeploymentInstance{}, instance.ID).Error
```
Comment (teleivo) [inline pkg/instance/repository.go:26]: Can this also return a `gorm.ErrRecordNotFound` I see you are checking for it below

File: pkg/instance/service.go [hunk 9]
```diff
@@ -313,6 +315,63 @@ func (s service) deployDeploymentInstance(token string, instance *model.Deployme
...
+
+	var errs error
+	for _, instance := range instances {
+		err := s.destroyDeploymentInstance(instance)
+		if err != nil {
+			errs = errors.Join(errs, fmt.Errorf("failed to destroy instance(%s) %q: %v", instance.StackName, instance.Name, err))
+		}
+	}
+
+	if errs != nil {
```
Comment (teleivo) [inline pkg/instance/service.go:338]: What if we succeeded in destroying only some instances, should we delete these from the DB? 🤔 Imagine that case and a user retrying, would destroying the same instances that were destroyed before work again? Hence can a user retry and actually get all instances destroyed and removed from the DB in that case.

**Review by teleivo — APPROVED** [review]
